### ISSUE TITLE
Underscore in #normalized_host is replaced with %5F

### ISSUE THREAD

**Issue by scarfacedeb** [sporkmonger/addressable#233]
According to [RFC 3986](http://tools.ietf.org/html/rfc3986), `_` shouldn't be replaced with `%5F` in the **host**. (it also applies to `~`)

Concretely, host may contain:

```
  unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
```

And in Chapter 2.3 about unreserved characters, it's said that:

```
   For consistency, percent-encoded octets in the ranges of ALPHA
   (%41-%5A and %61-%7A), DIGIT (%30-%39), hyphen (%2D), period (%2E),
   underscore (%5F), or tilde (%7E) should not be created by URI
   producers and, when found in a URI, should be decoded to their
   corresponding unreserved characters by URI normalizers.
```

Regarding the **normalization**, in Chapter 6.2.2.2. it's advised against replacing the unreserved characters with percent-encoded octets:

```
6.2.2.2.  Percent-Encoding Normalization

   The percent-encoding mechanism (Section 2.1) is a frequent source of
   variance among otherwise identical URIs.  In addition to the case
   normalization issue noted above, some URI producers percent-encode
   octets that do not require percent-encoding, resulting in URIs that
   are equivalent to their non-encoded counterparts.  These URIs should
   be normalized by decoding any percent-encoded octet that corresponds
   to an unreserved character, as described in Section 2.3.
```
## Steps to reproduce

```
Addressable::URI.parse("http://sms_service.example.com").normalized_host
=> "sms%5Fservice.example.com"
```

Default ruby's implementation of `URI` handles it correctly (ver. 2.3.0):

```
URI.parse("http://sms_service.example.com").normalize.host
=> "sms_service.example.com"
```

**Comment by 907th** [comment]
Note:

Only allowed characters in DNS names are `[a-zA-Z0-9\-]`.

From [RFC 3986](http://tools.ietf.org/html/rfc3986#page-20):
> The most common name registry mechanism is the Domain Name System (DNS). Such a name consists of a sequence of domain labels separated by ".", each domain label starting and ending with an alphanumeric character and possibly also containing "-" characters.

Also, see [RFC 1034](http://tools.ietf.org/html/rfc1034#section-3.5) (DOMAIN NAMES - CONCEPTS AND FACILITIES).

So, strongly speaking, DNS name like `sms_service.example.co` is not allowed.

**Comment by scarfacedeb** [comment]
@907th you're right.

But replacing `_` with the percent-encoded value is also the incorrect behavior.

**Comment by sporkmonger** [comment]
Worth reading my comment here on this issue for some context:
https://github.com/sporkmonger/addressable/issues/161#issuecomment-87000114

I 100% agree w/ both of you and I'll look to add some new tests to address this.

**Comment by sporkmonger** [comment]
Fixed in f032606
